packages/vhd-lib/src/_vhdStreamGenerator.js

```diff
@@ -0,0 +1,130 @@
+import { BLOCK_UNUSED, FOOTER_SIZE, HEADER_SIZE, SECTOR_SIZE } from './_constants'
+import { readChunk } from '@vates/read-chunk'
+import assert from 'assert'
+import { unpackFooter, unpackHeader, computeBlockBitmapSize } from './Vhd/_utils'
+
+const cappedBufferConcat = (buffers, maxSize) => {
+  let buffer = Buffer.concat(buffers)
+  if (buffer.length > maxSize) {
+    buffer = buffer.slice(buffer.length - maxSize)
+  }
+  return buffer
+}
+
+export async function* vhdStreamGenerator(stream) {
+  let bytesRead = 0
+
+  // handle empty space between elements
+  // ensure we read stream in order
+  async function read(offset, size) {
+    assert(bytesRead <= offset, `offset is ${offset} but we already read ${bytesRead} bytes`)
+    if (bytesRead < offset) {
+      // empty spaces
+      await read(bytesRead, offset - bytesRead)
+    }
+    const buf = await readChunk(stream, size)
+    assert.strictEqual(buf.length, size, `read ${buf.length} instead of ${size}`)
+    bytesRead += size
+    return buf
+  }
+
+  const bufFooter = await read(0, FOOTER_SIZE)
+
+  const footer = unpackFooter(bufFooter)
+  yield { type: 'footer', footer, offset: 0 }
+
+  const bufHeader = await read(FOOTER_SIZE, HEADER_SIZE)
+  const header = unpackHeader(bufHeader, footer)
+
+  yield { type: 'header', header, offset: SECTOR_SIZE }
+  const blockSize = header.blockSize
+  assert.strictEqual(blockSize % SECTOR_SIZE, 0)
+
+  const blockBitmapSize = computeBlockBitmapSize(blockSize)
+  const blockAndBitmapSize = blockBitmapSize + blockSize
+
+  const index = []
+
+  for (const parentLocatorId in header.parentLocatorEntry) {
+    const parentLocatorEntry = header.parentLocatorEntry[parentLocatorId]
+    // empty parent locator entry, does not exist in the content
+    if (parentLocatorEntry.platformDataSpace === 0) {
+      continue
+    }
+    index.push({
+      ...parentLocatorEntry,
+      type: 'parentLocator',
+      offset: parentLocatorEntry.platformDataOffset,
+      size: parentLocatorEntry.platformDataLength,
+      id: parentLocatorId,
+    })
+  }
+
+  const batOffset = header.tableOffset
+  const batSize = Math.max(1, Math.ceil((header.maxTableEntries * 4) / SECTOR_SIZE)) * SECTOR_SIZE
+
+  index.push({
+    type: 'bat',
+    offset: batOffset,
+    size: batSize,
+  })
+
+  // sometimes some parent locator are before the BAT
+  index.sort((a, b) => a.offset - b.offset)
+  while (index.length > 0) {
+    const item = index.shift()
+    const buffer = await read(item.offset, item.size)
+    if (item.type === 'bat') {
+      // found the BAT : read it and ad block to index
+      const bat = Uint32Array(header.maxTableEntries)
```

Comment:
You can unpack the BAT directly and use it like a simple array:

```js
const bat = new Uint32Array(buffer.buffer, buffer.bytesOffset, buffer.length / Uint32Array.BYTES_PER_ELEMENT)

const blockOffset = bat[blockId]
```

Reply:
I recompute the block offset , so the resulting BAT will be different than the source BAT

Reply:
Endianness is not guaranteed with `Uint32Array`, let's forget my comment.